File: backend/app/models/company_billing.py
from datetime import date, datetime

from sqlalchemy import Date, DateTime, ForeignKey, Numeric, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.master_database import MasterBase


class CompanyBilling(MasterBase):
    __tablename__ = "company_billings"

    id: Mapped[int] = mapped_column(primary_key=True)
    company_id: Mapped[int] = mapped_column(ForeignKey("companies.id"), index=True)
    reference_month: Mapped[str] = mapped_column(String(7), index=True)
    due_date: Mapped[date] = mapped_column(Date, index=True)
    amount: Mapped[float] = mapped_column(Numeric(12, 2), nullable=False)
    # Encargos ficam separados do principal para preservar o valor contratado.
    interest_amount: Mapped[float] = mapped_column(Numeric(12, 2), default=0, nullable=False)
    late_fee_amount: Mapped[float] = mapped_column(Numeric(12, 2), default=0, nullable=False)
    monetary_correction_amount: Mapped[float] = mapped_column(Numeric(12, 2), default=0, nullable=False)
    waived_amount: Mapped[float] = mapped_column(Numeric(12, 2), default=0, nullable=False)
    waived_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    waived_by: Mapped[str | None] = mapped_column(String(120))
    waiver_reason: Mapped[str | None] = mapped_column(Text)
    payment_method: Mapped[str | None] = mapped_column(String(40))
    status: Mapped[str] = mapped_column(String(20), default="pending", index=True)
    paid_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    paid_amount: Mapped[float | None] = mapped_column(Numeric(12, 2))
    mercado_pago_payment_id: Mapped[str | None] = mapped_column(String(80), index=True)
    mercado_pago_status: Mapped[str | None] = mapped_column(String(40))
    mercado_pago_payer_name: Mapped[str | None] = mapped_column(String(180))
    mercado_pago_payer_email: Mapped[str | None] = mapped_column(String(180))
    mercado_pago_payer_document: Mapped[str | None] = mapped_column(String(80))
    mercado_pago_external_reference: Mapped[str | None] = mapped_column(String(120), index=True)
    mercado_pago_idempotency_key: Mapped[str | None] = mapped_column(String(80))
    pix_qr_code: Mapped[str | None] = mapped_column(Text)
    pix_qr_code_base64: Mapped[str | None] = mapped_column(Text)
    pix_ticket_url: Mapped[str | None] = mapped_column(Text)
    notes: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
    )

    company = relationship("Company")

File: admin_app/admin_flutter/lib/models/company_billing.dart
class CompanyBilling {
  const CompanyBilling({
    required this.id,
    required this.companyId,
    required this.companyCode,
    required this.companyName,
    required this.referenceMonth,
    required this.dueDate,
    required this.amount,
    this.interestAmount = 0,
    this.lateFeeAmount = 0,
    this.monetaryCorrectionAmount = 0,
    this.waivedAmount = 0,
    this.totalDue,
    this.paymentMethod,
    required this.status,
    this.paidAt,
    this.paidAmount,
    this.mercadoPagoPaymentId,
    this.mercadoPagoStatus,
    this.mercadoPagoPayerName,
    this.mercadoPagoPayerEmail,
    this.mercadoPagoPayerDocument,
    this.pixQrCode,
    this.pixQrCodeBase64,
    this.pixTicketUrl,
    this.notes,
  });

  final int id;
  final int companyId;
  final String companyCode;
  final String companyName;
  final String referenceMonth;
  final DateTime dueDate;
  final double amount;
  final double interestAmount;
  final double lateFeeAmount;
  final double monetaryCorrectionAmount;
  final double waivedAmount;
  final double? totalDue;
  final String? paymentMethod;
  final String status;
  final DateTime? paidAt;
  final double? paidAmount;
  final String? mercadoPagoPaymentId;
  final String? mercadoPagoStatus;
  final String? mercadoPagoPayerName;
  final String? mercadoPagoPayerEmail;
  final String? mercadoPagoPayerDocument;
  final String? pixQrCode;
  final String? pixQrCodeBase64;
  final String? pixTicketUrl;
  final String? notes;

  bool get isOverdue =>
      status == 'pending' &&
      DateTime(dueDate.year, dueDate.month, dueDate.day).isBefore(
        DateTime(DateTime.now().year, DateTime.now().month, DateTime.now().day),
      );

  factory CompanyBilling.fromJson(Map<String, dynamic> json) {
    return CompanyBilling(
      id: json['id'] as int,
      companyId: json['company_id'] as int,
      companyCode: json['company_code'] as String,
      companyName: json['company_name'] as String,
      referenceMonth: json['reference_month'] as String,
      dueDate: DateTime.parse(json['due_date'] as String),
      amount: double.parse(json['amount'].toString()),
      interestAmount: double.tryParse('${json['interest_amount'] ?? 0}') ?? 0,
      lateFeeAmount: double.tryParse('${json['late_fee_amount'] ?? 0}') ?? 0,
      monetaryCorrectionAmount:
          double.tryParse('${json['monetary_correction_amount'] ?? 0}') ?? 0,
      waivedAmount: double.tryParse('${json['waived_amount'] ?? 0}') ?? 0,
      totalDue: json['total_due'] == null
          ? null
          : double.tryParse('${json['total_due']}'),
      paymentMethod: json['payment_method']?.toString(),
      status: json['status']?.toString() ?? 'pending',
      paidAt: json['paid_at'] == null
          ? null
          : DateTime.parse(json['paid_at'] as String),
      paidAmount: json['paid_amount'] == null
          ? null
          : double.parse(json['paid_amount'].toString()),
      mercadoPagoPaymentId: json['mercado_pago_payment_id']?.toString(),
      mercadoPagoStatus: json['mercado_pago_status']?.toString(),
      mercadoPagoPayerName: json['mercado_pago_payer_name']?.toString(),
      mercadoPagoPayerEmail: json['mercado_pago_payer_email']?.toString(),
      mercadoPagoPayerDocument: json['mercado_pago_payer_document']?.toString(),
      pixQrCode: json['pix_qr_code']?.toString(),
      pixQrCodeBase64: json['pix_qr_code_base64']?.toString(),
      pixTicketUrl: json['pix_ticket_url']?.toString(),
      notes: json['notes']?.toString(),
    );
  }
}

class CompanyBillingCreate {
  const CompanyBillingCreate({
    required this.companyId,
    required this.referenceMonth,
    required this.dueDate,
    required this.amount,
    this.paymentMethod,
    this.notes,
  });

  final int companyId;
  final String referenceMonth;
  final DateTime dueDate;
  final double amount;
  final String? paymentMethod;
  final String? notes;

  Map<String, dynamic> toJson() {
    return {
      'company_id': companyId,
      'reference_month': referenceMonth,
      'due_date': _dateOnly(dueDate),
      'amount': amount,
      'payment_method': paymentMethod,
      'notes': notes,
    };
  }
}

class CompanyBillingUpdate {
  const CompanyBillingUpdate({
    this.dueDate,
    this.amount,
    this.paymentMethod,
    this.status,
    this.notes,
  });

  final DateTime? dueDate;
  final double? amount;
  final String? paymentMethod;
  final String? status;
  final String? notes;

  Map<String, dynamic> toJson() {
    return {
      if (dueDate != null) 'due_date': _dateOnly(dueDate!),
      if (amount != null) 'amount': amount,
      if (paymentMethod != null) 'payment_method': paymentMethod,
      if (status != null) 'status': status,
      if (notes != null) 'notes': notes,
    };
  }
}

String _dateOnly(DateTime date) =>
    '${date.year.toString().padLeft(4, '0')}-${date.month.toString().padLeft(2, '0')}-${date.day.toString().padLeft(2, '0')}';
